overlord/snapstate/snapstate.go

```diff
@@ -1969,17 +1969,129 @@ func updateManyFiltered(ctx context.Context, st *state.State, names []string, re
 		return nil, nil, err
 	}
 
-	updated, updateTss, err := doUpdate(ctx, st, names, toUpdate, params, userID, flags, nil, deviceCtx, fromChange)
-	if err != nil {
-		return nil, nil, err
+	// if we're on classic with a kernel/gadget, split refreshes with essential
+	// snaps and apps so that the apps don't have to wait for a reboot
+	essential, nonEssential := splitEssentialUpdates(deviceCtx, toUpdate, names)
+
+	// TODO: support this for auto-refresh
+	if deviceCtx.IsCoreBoot() && release.OnClassic && !flags.IsAutoRefresh &&
```

Comment:
The code added here maybe could go to this own function, I think it would help with readability here.

Reply:
I've refactored this logic a bit so the next installPathMany PR is simpler. I think the new version makes this a bit more readable as well